File: EngPractice/Service/ChatService.cs
using EngPractice.Domain;
using Newtonsoft.Json;
using System.Text;
using System.Text.RegularExpressions;

namespace EngPractice.Service
{
    public class ChatService
    {
        private readonly HttpClient _httpClient;

        public ChatService(HttpClient httpClient)
        {
            _httpClient = httpClient;
        }

        public async Task<ChatResponse> GenerateAnswer(
            Conversation conversation,
            string username,
            string gender,
            int age,
            EnglishLevel englishLevel,
            string apiKey,
            bool enableReasoning,
            bool enableSearching)
        {
            // Kiểm tra API key
            if (!await HealthcheckService.Healthcheck(apiKey))
            {
                throw new Exception("API Key không hợp lệ.");
            }

            if (string.IsNullOrEmpty(apiKey))
            {
                throw new Exception("API Key không được để trống.");
            }

            var systemInstruction = enableReasoning
                ? Instructions.GetReasoningInstruction(username, gender, age, englishLevel)
                : enableSearching
                    ? Instructions.GetSearchingInstruction(username, gender, age, englishLevel)
                    : Instructions.GetBasicInstruction(username, gender, age, englishLevel);

            var requestBody = new
            {
                contents = new[]
                {
                    new
                    {
                        role = "user",
                        parts = new[] { new { text = conversation.Question } }
                    }
                },
                system_instruction = new { parts = new[] { new { text = systemInstruction } } },
                generationConfig = new
                {
                    maxOutputTokens = 1000,
                    temperature = enableSearching ? 0.3 : 1.0
                }
            };

            var requestContent = new StringContent(
                JsonConvert.SerializeObject(requestBody),
                Encoding.UTF8,
                "application/json");

            var response = await _httpClient.PostAsync(
                $"https://generativelanguage.googleapis.com/v1beta/models/gemini-1.5-flash:generateContent?key={apiKey}",
                requestContent);

            response.EnsureSuccessStatusCode();

            var responseBody = await response.Content.ReadAsStringAsync();
            Console.WriteLine($"Gemini response: {responseBody}");
            Console.WriteLine($"User question: {conversation.Question}");

            var geminiResponse = JsonConvert.DeserializeObject<dynamic>(responseBody);
            string message = geminiResponse.candidates[0].content.parts[0].text;

            Console.WriteLine($"Message before processing: {message}");

            // Xử lý Markdown nếu có
            if (message.StartsWith("```json"))
            {
                var jsonMatch = Regex.Match(message, @"```json\n([\s\S]*?)\n```");
                if (jsonMatch.Success)
                {
                    message = jsonMatch.Groups[1].Value;
                    Console.WriteLine($"Extracted JSON: {message}");
                }
                else
                {
                    throw new Exception($"Không thể trích xuất JSON từ Markdown: {message}");
                }
            }

            // Thử deserialize JSON
            ChatResponse chatResponse;
            try
            {
                chatResponse = JsonConvert.DeserializeObject<ChatResponse>(message);
            }
            catch (JsonException jsonEx)
            {
                Console.WriteLine($"JSON deserialization error: {jsonEx.Message}");
                // Xử lý như văn bản thuần túy nếu không phải JSON
                chatResponse = new ChatResponse
                {
                    MessageInMarkdown = message,
                    Suggestions = new List<string>
                    {
                        "Bạn muốn học từ vựng về chủ đề nào?",
                        "Chúng ta có thể luyện ngữ pháp cơ bản không?"
                    }
                };
            }

            if (enableSearching)
            {
                // Xử lý nguồn và gợi ý tìm kiếm nếu cần
            }

            return chatResponse;
        }
    }
}

File: backend-nestjs/src/chat/chat.service.ts
import { Injectable } from '@nestjs/common';
import { GoogleGenAI } from '@google/genai';
import { CreateChatDto } from './dto/create-chat.dto';
import { ChatPrompts } from './chat.constants';

@Injectable()
export class ChatService {
  async generateAnswer(dto: CreateChatDto) {
    try {
      const ai = new GoogleGenAI({ apiKey: dto.geminiApiKey });
      const systemInstructionText = ChatPrompts.getBasicInstruction(
        dto.username,
        dto.gender,
        dto.age,
        dto.englishLevel,
      );

      const contents = [
        ...dto.conversation.chatHistory
          .filter((msg) => msg.message && msg.message !== 'string')
          .map((msg) => ({
            role: msg.fromUser ? 'user' : 'model',
            parts: [{ text: msg.message }],
          })),
        { role: 'user', parts: [{ text: dto.conversation.question }] },
      ];

      const response = await ai.models.generateContent({
        model: 'gemini-2.5-flash',
        contents: contents,
        config: {
          systemInstruction: systemInstructionText,
          responseMimeType: 'application/json',
          temperature: 0.5,
        },
      });

      const text = response.text || 'Có lỗi xảy ra ở phía máy chủ, vui lòng thử lại sau.';

      return this.parseJsonResponse(text);
    } catch (error) {
      console.error('Chat Service Error:', error);
      if (error.status) {
        console.error(`Status: ${error.status}, Message: ${error.message}`);
      }

      return {
        MessageInMarkdown:
          'Xin lỗi, hệ thống đang gặp sự cố kết nối. Vui lòng thử lại sau.',
        Suggestions: [],
      };
    }
  }

  private parseJsonResponse(text: string | null) {
    if (!text) return { MessageInMarkdown: '', Suggestions: [] };
    try {
      const cleaned = text.replace(/^```json\s*/, '').replace(/\s*```$/, '');
      return JSON.parse(cleaned);
    } catch (e) {
      return {
        MessageInMarkdown: text,
        Suggestions: [],
      };
    }
  }
}
